layout: Make `transform-style: preserve-3d` establish a containing block for all descendants

HastD opened the pull request:
This makes `transform-style: preserve-3d` establish a containing block for all descendants, as specified here:
<https://drafts.csswg.org/css-transforms-2/#transform-style-property>

(This is my first time contributing to Servo; I think I followed the instructions properly but let me know if there's anything I need to fix.)

-----

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #35801 and #35824
- [x] There are tests for these changes

HastD committed:
layout: Fix behavior of `transform-style: preserve-3d`

This makes `transform-style: preserve-3d` establish a containing block
for all descendants, as specified here:
<https://drafts.csswg.org/css-transforms-2/#transform-style-property>

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#51167) with upstreamable changes.

Loirooriol commented on components/layout_2020/style_ext.rs:
```diff
@@ -688,6 +688,13 @@ impl ComputedValuesExt for ComputedValues {
             return true;
         }
 
+        // See <https://drafts.csswg.org/css-transforms-2/#transform-style-property>.
+        if self.get_box().transform_style == ComputedTransformStyle::Preserve3d ||
+            self.overrides_transform_style()
+        {
+            return true;
+        }
```
Ah, this needs to check if the element is transformable, see my comment in https://github.com/servo/servo/pull/35787#discussion_r1983608974

HastD replied:
I see, thanks. Should that helper method you mentioned be added in a different PR or could I go ahead and add it as part of this one?

Loirooriol replied:
You can add it here.

HastD committed:
layout: Check for transformable elements

Adds a new `is_transformable` helper method and use this in several other
methods, including the methods for whether the fragment establishes a
new stacking context or a containing block for all descendants.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51167).

Loirooriol commented on components/layout_2020/style_ext.rs:
```diff
@@ -688,6 +693,14 @@ impl ComputedValuesExt for ComputedValues {
+        // See <https://drafts.csswg.org/css-transforms-2/#transform-style-property>.
+        if self.is_transformable(fragment_flags) &&
+            (self.get_box().transform_style == ComputedTransformStyle::Preserve3d ||
+                self.overrides_transform_style())
```
`/_mozilla/css/opacity_stacking_context_a.html` is failing because this change makes `opacity: 0.5` establish a containing block.

I think you shouldn't check `self.overrides_transform_style()`, see #35824

HastD replied:
Fixed (in both `establishes_stacking_context` and `establishes_containing_block_for_all_descendants`). I confirmed that this makes the `opacity_stacking_context_a.html` test pass.

HastD committed:
Use generic green square reference for reftest.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
layout: Fix stacking context & containing block checks.

Only the computed value of `transform-style` should be used to determine
whether the element establishes a stacking context and/or a containing
block, not the used value.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
Merge branch 'servo:main' into preserve3d-containing-block

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

HastD commented on tests/wpt/tests/css/css-transforms/preserve3d-containing-block.html:
```diff
@@ -0,0 +1,12 @@
+<meta charset="utf-8">
+<title>CSS test: "transform-style: preserve-3d" should establish a containing block for all descendants</title>
+<link rel="help" href="https://drafts.csswg.org/css-transforms-2/#transform-style-property">
+<link rel="match" href="preserve3d-containing-block-ref.html">
```
Done.

Loirooriol reviewed (approved):
Not sure if the merge queue will work, CI seems broken

servo-wpt-sync commented:
✍ Updated existing upstream WPT pull request (web-platform-tests/wpt#51167) title and body.

Loirooriol commented:
`/css/css-masking/clip/clip-no-stacking-context.html` is now passing as a result of fixing  #35824

HastD committed:
Update clip-no-stacking-context test expectation to pass.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
Merge branch 'main' into preserve3d-containing-block

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#51167).
